x-pack/legacy/plugins/apm/server/lib/helpers/es_client.ts

```diff
@@ -165,3 +166,18 @@ export function getESClient(
     }
   };
 }
+
+export type SearchClient = ReturnType<typeof getSearchClient>;
+export function getSearchClient(callCluster: CallCluster) {
```

Comment:
Shouldn't this be called `getInternalEsClient` to be consistent with `getESClient` and `getInternalSavedObjectsClient`?

```suggestion
export function getInternalEsClient(callCluster: CallCluster) {
```

Reply:
Actually, `getESClient` only relies on `request` or debugging purposes. Could you use that instead of making a new client handler?

Reply:
After playing around with this for a bit I came up with this:
```ts
function getESClient(
  savedObjectsClient: SavedObjectsClientContract,
  uiSetttingsClient: UiSettingsClient,
  esClient:
    | IScopedClusterClient['callAsCurrentUser']
    | IScopedClusterClient['callAsInternalUser'],
  config: APMConfig,
  onRequest: (endpoint: string, params: Record<any, any>) => void
) 
```

All clients are explicitly passed in. This way makes its up to the consumer to decide whether the internal user or the request user should be used. I've replaced the `request` argument with an `onRequest` callback that can be used to log elasticsearch queries.

Reply:
This will also affect any downstream dependency that depends on `context` like `getParamsForSearchRequest`:

**Before**
```ts
async function getParamsForSearchRequest(
  context: APMRequestHandlerContext,
  params: ESSearchRequest,
  apmOptions?: APMOptions
) {
```

**After**
```ts
async function getParamsForSearchRequest(
  savedObjectsClient: SavedObjectsClientContract,
  uiSettingsClient: IUiSettingsClient,
  config: APMConfig,
  params: ESSearchRequest,
  apmOptions?: APMOptions
) {
```

@dgieselaar any opinions on something like this?

Reply:
Full example of `getEsClient`
```ts
export function getESClient(
  apmIndices: ApmIndicesConfig,
  uiSettingsClient: IUiSettingsClient,
  esClient:
    | IScopedClusterClient['callAsCurrentUser']
    | IScopedClusterClient['callAsInternalUser'],
  onRequest: (endpoint: string, params: Record<any, any>) => void = () => {}
) {
  return {
    search: async <
      TDocument = unknown,
      TSearchRequest extends ESSearchRequest = {}
    >(
      params: TSearchRequest,
      apmOptions?: APMOptions
    ): Promise<ESSearchResponse<TDocument, TSearchRequest>> => {
      const nextParams = await getParamsForSearchRequest(
        apmIndices,
        uiSettingsClient,
        params,
        apmOptions
      );

      onRequest('search', nextParams);

      return (esClient('search', nextParams) as unknown) as Promise<
        ESSearchResponse<TDocument, TSearchRequest>
      >;
    },
    index: <Body>(params: APMIndexDocumentParams<Body>) => {
      onRequest('index', params);
      return esClient('index', params);
    },
    delete: (params: IndicesDeleteParams) => {
      onRequest('delete', params);
      return esClient('delete', params);
    },
    indicesCreate: (params: IndicesCreateParams) => {
      onRequest('indices.create', params);
      return esClient('indices.create', params);
    }
  };
}
```

Reply:
OK i made this change in the latest version, the only other change was in setup_request, so it wasn't too invasive